public/app/core/reducers/createCollection.ts

```diff
@@ -0,0 +1,61 @@
+import { PayloadAction } from '@reduxjs/toolkit';
+import { StoreState } from '../../types';
+import { Reducer } from 'redux';
+
+export interface CollectionReducerState<InstanceState extends {}> {
+  [key: string]: InstanceState;
+}
+
+interface CollectionAction {
+  id: string;
+  action: PayloadAction<any>;
+}
+export const COLLECTION_UNKNOWN_ID = 'unknown-id';
+const COLLECTION_ACTION_PREFIX = 'collectionAction::';
+export const toCollectionAction = (action: PayloadAction<any>, id: string): PayloadAction<CollectionAction> => ({
+  type: `${COLLECTION_ACTION_PREFIX}${action.type}`, // makes it easier to debug in Redux dev tools
+  payload: { id, action },
+});
+
+export const createCollection = <InstanceState extends {}>(args: {
+  instanceReducer: Reducer<InstanceState>;
+  stateSelector: (state: StoreState) => CollectionReducerState<InstanceState>;
+}) => {
+  const { instanceReducer, stateSelector } = args;
+
+  // there might be a better redux toolkit way to create HOC reducer but I couldn't find anything
+  const reducer = (
+    state: CollectionReducerState<InstanceState> = {},
+    collectionAction: PayloadAction<CollectionAction>
+  ): CollectionReducerState<InstanceState> => {
+    if (collectionAction.type.indexOf(COLLECTION_ACTION_PREFIX) !== 0) {
+      return state;
+    }
+
+    const { id, action } = collectionAction.payload;
+    const collectionId = id ?? COLLECTION_UNKNOWN_ID;
+
+    const oldState = state[collectionId];
+    const newState = instanceReducer(oldState, action);
+
+    return {
+      ...state,
+      [collectionId]: {
+        ...oldState,
+        ...newState,
+      },
+    };
+  };
+
+  const selector = (state: StoreState, id: string): InstanceState => {
+    const collectionId = id ?? COLLECTION_UNKNOWN_ID;
```

Comment:
So there is a potential issue here if you navigate between dashboards without uid. For instance, navigating from home dashboard and new dashboard will cause a memory leak. Because the first dashboard without uid (home dashboard in this example) will not be cleaned up. I'm looking into this.

Reply:
Resolved with 5befa248261e6f704b88716d0244108b9dbde8f9